share/ipld/proof.go

```diff
@@ -15,37 +15,60 @@ type Proof struct {
 // proofCollector collects proof nodes' CIDs for the construction of a shares inclusion validation
 // nmt.Proof.
 type proofCollector struct {
-	left, right []cid.Cid
+	left, right []node
+}
+
+type node struct {
+	cid   cid.Cid
+	depth int
 }
 
 func newProofCollector(maxShares int) *proofCollector {
 	// maximum possible amount of required proofs from each side is equal to tree height.
-	height := int(math.Log2(float64(maxShares)))
+	height := int(math.Log2(float64(maxShares))) + 1
 	return &proofCollector{
-		left:  make([]cid.Cid, 0, height),
-		right: make([]cid.Cid, 0, height),
+		left:  make([]node, height),
+		right: make([]node, height),
 	}
 }
 
-func (c *proofCollector) addLeft(node cid.Cid) {
-	c.left = append(c.left, node)
+func (c *proofCollector) addLeft(cid cid.Cid, depth int) {
+	c.left[depth] = node{
```

Comment:
why do you need to put in depth index if you're gonna sort anyway?

Reply:
Depth is needed to make inserts lock free. Not sure if that what you meant, but I've removed sorts, since slices already sorted.